src/main/java/org/jenkinsci/maven/plugins/hpi/RunMojo.java

```diff
@@ -605,93 +600,30 @@ private String loadVersion(InputStream is) throws IOException {
-                }
-            }
-        });
-    }
-
-    @Override
-    public void restartWebApp(boolean reconfigureScanner) throws Exception {
-        getLog().info("restarting "+webApp);
-        getLog().debug("Stopping webapp ...");
-        webApp.stop();
-        getLog().debug("Reconfiguring webapp ...");
-
-        checkPomConfiguration();
-        configureWebApplication();
-
-        // check if we need to reconfigure the scanner,
-        // which is if the pom changes
-        if (reconfigureScanner)
-        {
-            getLog().info("Reconfiguring scanner after change to pom.xml ...");
-            generateHpl(); // regenerate hpl if POM changes.
-
-            scanList.clear();
-            if (webApp.getDescriptor() != null)
-                scanList.add(new File(webApp.getDescriptor()));
-            if (webApp.getJettyEnvXml() != null)
-                scanList.add(new File(webApp.getJettyEnvXml()));
-            scanList.add(project.getFile());
-            if (webApp.getTestClasses() != null)
-                scanList.add(webApp.getTestClasses());
-            if (webApp.getClasses() != null)
-            scanList.add(webApp.getClasses());
-            scanList.addAll(webApp.getWebInfLib());
-            scanner.setScanDirs(scanList);
-        }
-
-        getLog().debug("Restarting webapp ...");
-        webApp.start();
-        getLog().info("Restart completed at " + new Date());
-    }
-
-    private void setUpScanList() {
-        scanList = new ArrayList<>();
-        scanList.add(getProject().getFile());
-        scanList.add(webAppFile);
-        scanList.add(new File(getProject().getBuild().getOutputDirectory()));
+        // this can only be done after MavenServerConnector.doStart() is called, which happens in AbstractWebAppMojo.startJetty()
+        // and this startScanner method is one of the few places that are run afterward.
+        HttpConfiguration hc = httpConnector.getConnectionFactory(HttpConnectionFactory.class).getHttpConfiguration();
+        hc.setHttpCompliance(HttpCompliance.RFC7230);
+        hc.setUriCompliance(UriCompliance.LEGACY);
+        // Use a bigger buffer, as Stapler traces can get pretty large on deeply nested URLs.
+        hc.setResponseHeaderSize(12 * 1024);
+
+        super.startScanner();
     }
 
     @Override
-    protected void startConsoleScanner() throws Exception {
-        if (consoleForceReload) {
-            getLog().info("Console reloading is ENABLED. Hit ENTER on the console to restart the context.");
-            consoleScanner = new ConsoleScanner(this);
-            consoleScanner.start();
+    protected boolean isPackagingSupported() {
+        if (!supportedPackagings.contains("hpi")) {
+            List<String> newSupportedPackagings = new ArrayList<>(supportedPackagings);
+            newSupportedPackagings.add("hpi");
+            supportedPackagings = List.copyOf(newSupportedPackagings);
         }
+        return super.isPackagingSupported();
     }
```

Comment:
I considered forcing consumers to add

```xml
              <supportedPackagings>
                <supportedPackaging>hpi</supportedPackaging>
              </supportedPackagings>
```

but felt it would be smoother if the default included `hpi`. I didn't remove `war` but I don't feel strongly about this.